### ISSUE TITLE
Last apiserver to shutdown doesn't remove its IP from k8s svc endpoints

### ISSUE THREAD

**Issue by shyamjvs** [kubernetes/kubernetes#115804]
### What happened?

As part of the apiserver's pre-shutdown sequence, we stop the lease-controller from renewing the apiserver's etcd lease anymore and delete the lease:

https://github.com/kubernetes/kubernetes/blob/125e38c0872521bec49351415cc9fb624b046659/pkg/controlplane/controller.go#L211-L215

As part of the `RemoveEndpoints` call we expect that after the lease is deleted, the apiserver's IP is also removed from the k8s service endpoints object:

https://github.com/kubernetes/kubernetes/blob/125e38c0872521bec49351415cc9fb624b046659/pkg/controlplane/reconcilers/lease.go#L315-L321

However, `doReconcile` isn't removing the IP if it's the last remaining one:

https://github.com/kubernetes/kubernetes/blob/125e38c0872521bec49351415cc9fb624b046659/pkg/controlplane/reconcilers/lease.go#L210-L215

The comment there is assuming that `doReconcile` is always called after renewing the lease, which is incorrect since it can also be called as part of the shutdown workflow.

### What did you expect to happen?

The IP of the apiserver (even though it's the last one remaining) should be removed from k8s svc endpoints on shutdown.

### How can we reproduce it (as minimally and precisely as possible)?

Shutdown the last apiserver of the cluster and you'll see logs like this:

```
I1031 17:34:49.517758      12 controller.go:181] Shutting down kubernetes service endpoint reconciler
E1031 17:34:49.534880      12 controller.go:184] no master IPs were listed in storage, refusing to erase all endpoints for the kubernetes service
```

And the audit logs also confirm that the IP isn't removed from k8s endpoints object.

#### Why is this an issue?

If the IP isn't removed from the k8s endpoints object, clients (using in-cluster API mode) will try to continue talking/connecting to that instance even after the apiserver is dead. There are at least two issues with this:
- Causes confusing `i/o timeout` and `connection refused` errors for clients, instead of a cleaner `no route to host` error
- If the apiserver pod is restarted by kubelet (let's say due to healthcheck failure), clients may start talking to it prematurely even without healthz/readyz passing. This is not a recommended behavior ([xref](https://kubernetes.io/docs/reference/using-api/health-checks/)) as a bunch post-startup hooks like informer sync, internal controllers, etc may not have initialized yet

/sig api-machinery

### Kubernetes version

All versions till HEAD.

### Cloud provider

N/A

**Comment by shyamjvs** [comment]
cc @wojtek-t (as I see you've made some recent edits around the lease-controller code)

If we agree removing the IP is the right behavior, I can send out a fix for this issue.
/assign

**Comment by aojea** [comment]
> If the IP isn't removed from the k8s endpoints object, clients (using in-cluster API mode) will try to continue talking/connecting to that instance even after the apiserver is dead

if there is no apiserver ... how do you read the IP from the Service?  ... it is impossible since there is no apiserver to serve it

**Comment by shyamjvs** [comment]
Kube-proxy watches the endpoints from the apiserver, and during the (graceful) shutdown sequence the watch is still able to deliver events fwis. There isn't a guarantee though the event is delivered to all clients before shutdown completes (best effort).

**Comment by aojea** [comment]
/triage accepted

that logic you pointed out is wrong as you correctly describe, and I think that you are right about the watchers, those connections will stay a bit longer , there is more detail and tests about the shutdown process in

https://github.com/kubernetes/kubernetes/blob/ebae41641faa832460d3587e1fd13b39d1134152/staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go#L123-L172

and you have an integration test here too

https://github.com/kubernetes/kubernetes/blob/ebae41641faa832460d3587e1fd13b39d1134152/test/integration/controlplane/graceful_shutdown_test.go

@shyamjvs are you going to PR it? please add coverage with tests to avoid regressions and tag me for the review

**Comment by shyamjvs** [comment]
Yes, I'll send out the fix (or find someone who wants to do it) by end of this week. Also gives some time for folks to chime in with any other concerns.

**Comment by cmssczy** [comment]
Do you think add a bool flag in `RemoveEndpoints` to indicate the controller that we are shutting down is a good idea?

**Comment by aojea** [comment]
yeah, we have that pattern in some places,

**Comment by jaidevmane** [comment]
Hi folks,
I will be working with @shyamjvs to get out a PR and testcases for the above.

**Comment by cmssczy** [comment]
@shyamjvs Have you started working on this problem?  I'd like to work on it if you haven't started.

**Comment by shyamjvs** [comment]
@cmssczy Thanks for offering to help. @jaidevmane reached out to me earlier and is a new contributor eager to help. Would like to have him take this one please.

**Comment by jaidevmane** [comment]
Hi folks,
I would no longer be able to contribute to this issue due to other commitments. Would request someone else to take it up.
@czybjtu Do have a look if interested.

**Comment by czybjtu** [comment]
> I think that you can add a `ready bool` and set it to true when the controller starts, and set to `false` when shutting down

`r.stopReconcilingCalled` will be set to true only when apiserver was shut down.
https://github.com/kubernetes/kubernetes/blob/8decaf3ae7f410ab3f3774f3895b9f3124b8a4c6/pkg/controlplane/reconcilers/lease.go#L323-L327

So I think we could use `stopReconcilingCalled` as a bool flag to indicate if we should remove the last master ip, instead of add a new one. WDYT? @aojea 
https://github.com/kubernetes/kubernetes/blob/8decaf3ae7f410ab3f3774f3895b9f3124b8a4c6/pkg/controlplane/reconcilers/lease.go#L146-L151

**Comment by aojea** [comment]
interesting, that bool is called under lock

https://github.com/kubernetes/kubernetes/blob/8decaf3ae7f410ab3f3774f3895b9f3124b8a4c6/pkg/controlplane/reconcilers/lease.go#L169-L184

and the reconcile process is locked, so it that changes during the reconcile process the internal loop will never see it.
If that changes before the process the internal loop is not going to be executed.

@czybjtu let's try to remove that lock and bool with an atomic.bool  and improve the shutdown process, we should not bail out soon if we want to update the lease ... send a PR with these things and we'll take it from there

**Comment by czybjtu** [comment]
I found that `ReconcileEndpoints` and `StopReconciling` function are called in different goroutines. If we remove the lock, will it cause a race condition problem?

**Comment by aojea** [comment]
remove the lock and replace the bool by `atomic.Bool`

**Comment by aojea** [comment]
I think that something like this should work, but feel free to suggest improvements, and please add a test in lease_test.go to cover this bug

```diff
diff --git a/pkg/controlplane/reconcilers/lease.go b/pkg/controlplane/reconcilers/lease.go
index 9911d6a4272..b4c8a05abf5 100644
--- a/pkg/controlplane/reconcilers/lease.go
+++ b/pkg/controlplane/reconcilers/lease.go
@@ -26,6 +26,7 @@ import (
        "net"
        "path"
        "sync"
+       "sync/atomic"
        "time"
 
        "k8s.io/klog/v2"
@@ -146,8 +147,7 @@ func NewLeases(config *storagebackend.ConfigForResource, baseKey string, leaseTi
 type leaseEndpointReconciler struct {
        epAdapter             EndpointsAdapter
        masterLeases          Leases
-       stopReconcilingCalled bool
-       reconcilingLock       sync.Mutex
+       stopReconcilingCalled atomic.Bool
 }
 
 // NewLeaseEndpointReconciler creates a new LeaseEndpoint reconciler
@@ -167,10 +167,7 @@ func NewLeaseEndpointReconciler(epAdapter EndpointsAdapter, masterLeases Leases)
 // different from the directory listing, and update the endpoints object
 // accordingly.
 func (r *leaseEndpointReconciler) ReconcileEndpoints(serviceName string, ip net.IP, endpointPorts []corev1.EndpointPort, reconcilePorts bool) error {
-       r.reconcilingLock.Lock()
-       defer r.reconcilingLock.Unlock()
-
-       if r.stopReconcilingCalled {
+       if r.stopReconcilingCalled.Load() {
                return nil
        }
 
@@ -184,6 +181,7 @@ func (r *leaseEndpointReconciler) ReconcileEndpoints(serviceName string, ip net.
        return r.doReconcile(serviceName, endpointPorts, reconcilePorts)
 }
 
+// doReconcile can be called from ReconcileEndpoints() or RemoveEndpoints()
 func (r *leaseEndpointReconciler) doReconcile(serviceName string, endpointPorts []corev1.EndpointPort, reconcilePorts bool) error {
        e, err := r.epAdapter.Get(corev1.NamespaceDefault, serviceName, metav1.GetOptions{})
        shouldCreate := false
@@ -192,6 +190,11 @@ func (r *leaseEndpointReconciler) doReconcile(serviceName string, endpointPorts
                        return err
                }
 
+               // there are no endpoints and we should stop reconciling
+               if !r.stopReconcilingCalled.Load() {
+                       return nil
+               }
+
                shouldCreate = true
                e = &corev1.Endpoints{
                        ObjectMeta: metav1.ObjectMeta{
@@ -210,7 +213,9 @@ func (r *leaseEndpointReconciler) doReconcile(serviceName string, endpointPorts
        // Since we just refreshed our own key, assume that zero endpoints
        // returned from storage indicates an issue or invalid state, and thus do
        // not update the endpoints list based on the result.
-       if len(masterIPs) == 0 {
+       // If the controller was ordered to stop and is this is the last apiserver
+       // we keep going to remove our endpoint before shutting down.
+       if len(masterIPs) == 0 && !r.stopReconcilingCalled.Load() {
                return fmt.Errorf("no master IPs were listed in storage, refusing to erase all endpoints for the kubernetes service")
        }
 
@@ -321,9 +326,7 @@ func (r *leaseEndpointReconciler) RemoveEndpoints(serviceName string, ip net.IP,
 }
 
 func (r *leaseEndpointReconciler) StopReconciling() {
-       r.reconcilingLock.Lock()
-       defer r.reconcilingLock.Unlock()
-       r.stopReconcilingCalled = true
+       r.stopReconcilingCalled.Store(true)
 }
 ```

**Comment by czybjtu** [comment]
Actually, leaseEndpointReconciler never deletes endpoints. It deletes lease resources. So it's not useful when we do like this.
```diff
 func (r *leaseEndpointReconciler) doReconcile(serviceName string, endpointPorts []corev1.EndpointPort, reconcilePorts bool) error {
        e, err := r.epAdapter.Get(corev1.NamespaceDefault, serviceName, metav1.GetOptions{})
        shouldCreate := false
@@ -192,6 +190,11 @@ func (r *leaseEndpointReconciler) doReconcile(serviceName string, endpointPorts
                        return err
                }
 
+               // there are no endpoints and we should stop reconciling
+               if r.stopReconcilingCalled.Load() {
+                       return nil
+               }
+
```
https://github.com/kubernetes/kubernetes/blob/742316ee2136b24b9edb0bb0d5bf69a6bad6f364/pkg/controlplane/reconcilers/lease.go#L316

The execution order of these two gorotinues is completely unpredictable. So I don't think we could use `atomic.Bool` instead of lock by add  some logic like`if r.stopReconcilingCalled.Load() { return ...}`

In other words, even if the `StopReconciling `  function deletes the endpoint, `ReconcileEndpoints` function may still add it back.

**Comment by aojea** [comment]
> Actually, leaseEndpointReconciler never deletes endpoints

is not about deleting the endpoint, is about deleting the IP address of the endpoint, the operations when deleting has to be always an Update to the Endpoint object

The doReconcile build the endpoint based on the existing leases in etcd.
If the apiserver is the last one, RemoveEndpoints first deletes the lease and calls doReconcile.
doReconcile doesn't update the Endpoints to indicate "there are no IPs", because of
```
 if len(masterIPs) == 0 { 
 	return fmt.Errorf("no master IPs were listed in storage, refusing to erase all endpoints for the kubernetes service") 
 } 
```

**Comment by czybjtu** [comment]
> If the apiserver is the last one, RemoveEndpoints first deletes the lease and calls doReconcile.
> doReconcile doesn't update the Endpoints to indicate "there are no IPs", because of

Both ReconcileEndpoints and RemoveEndpoints will call `doReconcile`.  Since they are executed in two separate goroutines, the order in which they execute the `doReconcile` function is unknown.  

I think RemoveEndpoints could do what we expected. 

`doReconcile` which called by `ReconcileEndpoints` function may add endpoint back which was just be deleted(by update method) , because `ReconcileEndpoints` executed periodically.

**Comment by aojea** [comment]
I see your point and you are right, we should make `doReconcile` and make it thread safe to avoid possible races.

> Finally we could still get an valid endpoint If all the event follow the timeline in the diagram, even though this is a small probability event and it will be reconciled in next period by ReconcileEndpoints.

However, this can always happen if the callers make a bad use of the methods, ReconcileEndpoints() create endpoints and RemoveEndpoints() remove them, we can only make their use more predictable

**Comment by rtheis** [comment]
FYI: I've opened https://github.com/kubernetes/kubernetes/issues/118777 which I think is related to the fix for this issue.
